tag-system/src/main/java/org/togetherjava/tjbot/tag/TagSystem.java

```diff
@@ -0,0 +1,66 @@
+package org.togetherjava.tjbot.tag;
+
+import org.jooq.Result;
+import org.togetherjava.tjbot.db.Database;
+import org.togetherjava.tjbot.db.generated.tables.Tags;
+import org.togetherjava.tjbot.db.generated.tables.records.TagsRecord;
+
+import java.util.*;
+
+public class TagSystem {
+    private final Database database;
+
+    public TagSystem(Database database) {
+        this.database = database;
+    }
+
+    public boolean exists(String tag) {
+        return database.readTransaction(ctx -> {
+            return Optional
+                .ofNullable(ctx.selectFrom(Tags.TAGS).where(Tags.TAGS.ID.eq(tag)).fetchOne())
+                .isPresent();
+        });
+    }
+
+    public void delete(String tag) {
+        database.writeTransaction(ctx -> {
+            ctx.deleteFrom(Tags.TAGS).where(Tags.TAGS.ID.eq(tag)).execute();
+        });
+    }
+
+    public void put(String tag, String content) {
+        database.writeTransaction(ctx -> {
+            TagsRecord tagsRecord = ctx.newRecord(Tags.TAGS).setId(tag).setContent(content);
+
+            if (tagsRecord.update() == 0) {
+                tagsRecord.insert();
+            }
+        });
+    }
+
+    public String get(String tag) {
+        return database.readTransaction(ctx -> {
+            return Optional
+                .ofNullable(ctx.selectFrom(Tags.TAGS).where(Tags.TAGS.ID.eq(tag)).fetchOne())
+                .map(TagsRecord::getContent)
+                .orElse(null);
+        });
+    }
+
+    public Map<String, String> retrieve() {
+        return database.readTransaction(ctx -> {
+            Result<TagsRecord> result = ctx.selectFrom(Tags.TAGS).fetch();
+            Map<String, String> out = new HashMap<>();
+
+            for (TagsRecord tagsRecord : result) {
+                out.put(tagsRecord.getId(), tagsRecord.getContent());
+            }
+
+            return out;
+        });
+    }
+
+    public Set<String> retrieveIds() {
+        return retrieve().keySet();
```

Comment:
if applicable, prefer instead:
```suggestion
        return Collections.unmodifiableSet(retrieve().keySet());
```

Reply:
done